pkg/experiment/metastore/metastore_state_poll_compaction_jobs.go

```diff
@@ -27,119 +26,122 @@ func (m *Metastore) PollCompactionJobs(_ context.Context, req *compactorv1.PollC
+		updatedJobs:        make([]string, 0),
+	}
+
+	for _, statusUpdate := range request.JobStatusUpdates {
+		job := m.findJob(statusUpdate.JobName)
+		if job == nil {
+			level.Error(m.logger).Log("msg", "error processing update for compaction job, job not found", "job", statusUpdate.JobName, "err", err)
+			continue
+		}
+		if !m.compactionJobQueue.isOwner(job.Name, statusUpdate.RaftLogIndex) {
+			level.Warn(m.logger).Log("msg", "job is not assigned to the worker", "job", statusUpdate.JobName, "raft_log_index", statusUpdate.RaftLogIndex)
+			continue
+		}
+		jobKey := tenantShard{
+			tenant: job.TenantId,
+			shard:  job.Shard,
+		}
+		level.Debug(m.logger).Log("msg", "processing status update for compaction job", "job", statusUpdate.JobName, "status", statusUpdate.Status)
+		switch statusUpdate.Status {
+		case compactorv1.CompactionStatus_COMPACTION_STATUS_SUCCESS:
+			m.compactionJobQueue.evict(job.Name, job.RaftLogIndex)
+			stateUpdate.deletedJobs[jobKey] = append(stateUpdate.deletedJobs[jobKey], job.Name)
+			m.compactionMetrics.completedJobs.WithLabelValues(
+				fmt.Sprint(job.Shard), job.TenantId, fmt.Sprint(job.CompactionLevel)).Inc()
+
+			for _, b := range statusUpdate.CompletedJob.Blocks {
+				m.getOrCreateShard(job.Shard).putSegment(b)
+				stateUpdate.newBlocks[job.Shard] = append(stateUpdate.newBlocks[job.Shard], b.Id)
+
+				if job := m.tryCreateJob(b, statusUpdate.RaftLogIndex); job != nil {
+					m.addCompactionJob(job)
+					stateUpdate.newJobs = append(stateUpdate.newJobs, job.Name)
+					m.compactionMetrics.addedJobs.WithLabelValues(
+						fmt.Sprint(job.Shard), job.TenantId, fmt.Sprint(job.CompactionLevel)).Inc()
+				} else {
+					m.addBlockToCompactionJobQueue(b)
 				}
-				return nil
-			})
-			if err != nil {
-				level.Error(m.logger).Log("msg", "error processing update for compaction job", "job", job.Name, "err", err)
-				continue
+				m.compactionMetrics.addedBlocks.WithLabelValues(
+					fmt.Sprint(job.Shard), job.TenantId, fmt.Sprint(job.CompactionLevel)).Inc()
+				stateUpdate.updatedBlockQueues[jobKey] = append(stateUpdate.updatedBlockQueues[jobKey], b.CompactionLevel)
 			}
-		}
-
-		if request.JobCapacity > 0 {
-			jResult.newJobAssignments, err = m.assignNewJobs(tx, int(request.JobCapacity), raft.Index, raft.AppendedAt.UnixNano())
-			if err != nil {
-				return err
+			for _, b := range job.Blocks {
+				m.getOrCreateShard(job.Shard).deleteSegment(b)
+				stateUpdate.deletedBlocks[job.Shard] = append(stateUpdate.deletedBlocks[job.Shard], b)
+				m.compactionMetrics.deletedBlocks.WithLabelValues(
+					fmt.Sprint(job.Shard), job.TenantId, fmt.Sprint(job.CompactionLevel)).Inc()
+			}
+		case compactorv1.CompactionStatus_COMPACTION_STATUS_IN_PROGRESS:
+			m.compactionJobQueue.update(statusUpdate.JobName, raftAppendedAtNanos, statusUpdate.RaftLogIndex)
+			stateUpdate.updatedJobs = append(stateUpdate.updatedJobs, job.Name)
+		case compactorv1.CompactionStatus_COMPACTION_STATUS_FAILURE:
```

Comment:
This is still missing a failure reason from the worker, I am considering adding it so that we can persist it and show it in tooling. However, after reaching max failures we currently delete the job to avoid accumulating jobs.

We could implement a job retention policy for this (we might need it anyway) but I would do that separately.